extensions-core/azure-extensions/src/main/java/org/apache/druid/storage/azure/AzureDataSegmentKiller.java

```diff
@@ -63,6 +69,79 @@ public AzureDataSegmentKiller(
     this.azureCloudBlobIterableFactory = azureCloudBlobIterableFactory;
   }
 
+  @Override
+  public void kill(List<DataSegment> segments) throws SegmentLoadingException
+  {
+    if (segments.isEmpty()) {
+      return;
+    }
+    if (segments.size() == 1) {
+      kill(segments.get(0));
+      return;
+    }
+
+    // create a list of keys to delete
+    Map<String, List<String>> containerToKeysToDelete = new HashMap<>();
+    for (DataSegment segment : segments) {
+      Map<String, Object> loadSpec = segment.getLoadSpec();
+      final String containerName = MapUtils.getString(loadSpec, "containerName");
+      final String blobPath = MapUtils.getString(loadSpec, "blobPath");
+      List<String> keysToDelete = containerToKeysToDelete.computeIfAbsent(
+          containerName,
+          k -> new ArrayList<>()
+      );
+      keysToDelete.add(blobPath);
+    }
+
+    boolean shouldThrowException = false;
+    for (Map.Entry<String, List<String>> containerToKeys : containerToKeysToDelete.entrySet()) {
+      shouldThrowException = deleteBlobKeys(containerToKeys.getValue(), containerToKeys.getKey());
```

Comment:
If shouldThrowException is true once, then it should be true always.

```suggestion
      shouldThrowException = shouldThrowException || deleteBlobKeys(containerToKeys.getValue(), containerToKeys.getKey());
```

Reply:
Actually my suggested pattern may not work, because we want to call `deleteBlobKeys` on every container, even if shouldThrowException is true. We probably want to follow the pattern in S3DataSegmentKiller.

```
hadException = deleteBlobKeys(containerToKeys.getValue(), containerToKeys.getKey());
if (hadException) {
  shouldThrowException = true;
}
```

Reply:
I believe if you do `true || evaluate_expression()` the jvm short circuits it and chooses not to evaluate the expression since it knows true || anything is true